src/algebra/algebra/tower.lean

```diff
@@ -298,6 +298,18 @@ le_antisymm (span_le.2 $ λ x hx, let ⟨p, q, hps, hqt, hpqx⟩ := set.mem_smul
   (λ _ _, add_mem _)
   (λ k x hx, smul_mem_span_smul' hs hx)
 
+/-- Turning `p : submodule S M` into an `R`-submodule gives the same module structure
+as turning it into a type and adding a module structure. -/
+@[simps]
+def restrict_scalars_equiv (p : submodule S A) : p.restrict_scalars R ≃ₗ[R] p :=
+-- Everything is automatically defeq except scalar multiplication.
+{ to_fun := λ x, x,
+  inv_fun := λ x, x,
+  map_smul' := λ c ⟨x, hx⟩, subtype.coe_injective $
+    by rw [← is_scalar_tower.algebra_map_smul S c (⟨x, hx⟩ : p), coe_smul, coe_smul,
+           is_scalar_tower.algebra_map_smul] ,
+  .. add_equiv.refl p }
```

Comment:
If you define this in the other direction the proof of `map_smul` is just `map_smul_of_tower`

Reply:
Although I'm also suspicious that `linear_equiv.refl` works fine here too!

Reply:
> If you define this in the other direction the proof of `map_smul` is just `map_smul_of_tower`

I don't quite see how: I tried:
```lean
linear_equiv.symm $
{ to_fun := λ x, x,
  inv_fun := λ x, x,
  map_smul' := λ c x, linear_map.map_smul_of_tower sorry c x,
  .. add_equiv.refl p }
```
with various choices for `sorry`, but can't find anything that works.

Reply:
> Although I'm also suspicious that `linear_equiv.refl` works fine here too!

Not really, `refl` puts the same `module` instance on either side, and here the main issue is that the smul is not defeq.